swap/peer.go

```diff
@@ -0,0 +1,195 @@
+	}
+
+	return nil
+}
+
+// handleChequeRequestMsg runs when a peer receives a `ChequeRequestMsg`
+// It is thus run by the debitor
+// So the debitor needs to:
+//   * check that it indeed owes to the requestor (if not, ignore message)
+//   * check serial number
+//   * check amount
+//   * if all is ok, issue the cheque
+func (sp *Peer) handleChequeRequestMsg(ctx context.Context, msg interface{}) (err error) {
+	// check we have indeed a negative balance with the peer
+	var req *ChequeRequestMsg
+	var ok bool
+	var peerBalance int64
+
+	// FIXME probably not needed
+	if req, ok = msg.(*ChequeRequestMsg); !ok {
+		return fmt.Errorf("Unexpected message type: %v", err)
+	}
+
+	peer := sp.ID()
+
+	sp.swap.lock.Lock()
+	defer sp.swap.lock.Unlock() //TODO: Do we really want to block so long?
+
+	if peerBalance, ok = sp.swap.balances[peer]; !ok {
+		return fmt.Errorf("No exchanges with peer: %v", peer)
+	}
+	// do we actually owe to the remote peer?
+	if peerBalance >= 0 {
+		return ErrDontOwe
+	}
+
+	// balance is negative, send a cheque
+	// TODO: merge with thresholds; need to check for threshold?
+	// if not, any negative balance will result in a cheque at this point
+
+	var cheque *Cheque
+
+	_ = sp.swap.loadCheque(peer)
+	lastCheque := sp.swap.cheques[peer]
+
+	amount := 0 - peerBalance
+
+	//TODO; need to have ChequeRequestMsg to contain last cheque and compare?
+	// emit cheque, send to peer
+	if lastCheque == nil {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: uint64(1),
+				Amount: uint64(amount),
+			},
+		}
+	} else {
+		cheque = &Cheque{
+			ChequeParams: ChequeParams{
+				Serial: lastCheque.Serial + 1,
+				Amount: lastCheque.Amount + uint64(amount),
+			},
+		}
+	}
+	cheque.ChequeParams.Timeout = defaultCashInDelay
+	cheque.ChequeParams.Contract = sp.swap.owner.Contract
+	cheque.Beneficiary = req.Beneficiary
+	cheque.Sig, err = sp.swap.signContent(cheque)
+	if err != nil {
+		return err
+	}
+
+	sp.swap.cheques[peer] = cheque
+	// TODO: what if there is an error here; is the cheque persisted?
+	err = sp.swap.stateStore.Put(peer.String()+"_cheques", &cheque)
+
+	// TODO: error handling might be quite more complex
+	if err != nil {
+		return err
+	}
+
+	emit := &EmitChequeMsg{
+		Cheque: cheque,
+	}
+
+	// TODO: reset balance here?
+	// if we don't, then multiple ChequeRequestMsg may be sent and multiple
+	// cheques will be generated
+	// If we do, then if something goes wrong and the remote does not reset the balance,
+	// we have issues as well.
+	// For now, reset the balance
+	sp.swap.resetBalance(peer)
+
+	return sp.Send(ctx, emit)
+}
+
+// handleEmitChequeMsg should be handled by the creditor when it receives
+// a cheque from a creditor
+// TODO: validate the contract address in the cheque to match the address given at handshake
+// TODO: this should not be blocking
+func (sp *Peer) handleEmitChequeMsg(ctx context.Context, msg interface{}) error {
+	chequeMsg, ok := msg.(*EmitChequeMsg)
+	if !ok {
```

Comment:
As explained in the other function, you can just modify the signature of this handler to take a concrete pointer type for `msg`, and not have to deal with `invalid message type` in all handlers, but only in `handleMsg`, where you select the appropriate handlers.

Reply:
Done.